File: src/credentialCache.ts
import { CredentialsError, readAccessToken, readFingerprint } from "./credentials";

/**
 * Injectable credential source. readFingerprint must never trigger a keychain
 * ACL prompt (metadata only); readSecret is the one call that may prompt.
 */
export interface CredentialSource {
  readFingerprint(): Promise<string | null>;
  readSecret(): Promise<string>;
}

const LOGIN_MESSAGE = "Could not read credentials. Log in with Claude Code.";
const DENIED_MESSAGE = "Keychain access denied. Use Refresh Now to try again.";
const EXPIRED_MESSAGE = "Authentication expired. Log in again.";

/**
 * Caches the access token so the secret store (macOS keychain / credentials
 * file) is read at most once per change. Every getToken() call does a
 * prompt-free fingerprint read; the secret is re-read only when the
 * fingerprint changes, so keychain ACL prompts are capped at roughly one per
 * Claude Code token rotation instead of one per poll.
 */
export class CredentialCache {
  private token: string | undefined;
  private fingerprint: string | null | undefined;
  private rejectedFingerprint: string | null | undefined;
  private rejectionMessage = EXPIRED_MESSAGE;

  constructor(
    private readonly source: CredentialSource = {
      readFingerprint,
      readSecret: readAccessToken,
    },
    private readonly env: NodeJS.ProcessEnv = process.env
  ) {}

  /**
   * Return the current access token.
   * force bypasses only a previous rejection (explicit user retry may
   * prompt once); a healthy cached token is returned without any secret read.
   */
  async getToken(force = false): Promise<string> {
    const envToken = this.env.CLAUDE_USAGE_TOKEN;
    if (envToken) {
      return envToken;
    }

    const fp = await this.source.readFingerprint();

    if (this.token !== undefined && fp === this.fingerprint) {
      return this.token;
    }

    if (fp === null) {
      this.token = undefined;
      this.fingerprint = null;
      throw new CredentialsError(LOGIN_MESSAGE);
    }

    if (!force && this.rejectedFingerprint !== undefined && fp === this.rejectedFingerprint) {
      throw new CredentialsError(this.rejectionMessage);
    }

    try {
      const token = await this.source.readSecret();
      this.token = token;
      this.fingerprint = fp;
      this.rejectedFingerprint = undefined;
      return token;
    } catch {
      this.token = undefined;
      this.rejectedFingerprint = fp;
      this.rejectionMessage = DENIED_MESSAGE;
      throw new CredentialsError(DENIED_MESSAGE);
    }
  }

  /** Call on HTTP 401/403: the cached token was rejected by the API. */
  invalidate(): void {
    this.token = undefined;
    this.rejectedFingerprint = this.fingerprint;
    this.rejectionMessage = EXPIRED_MESSAGE;
  }
}

export const defaultCache = new CredentialCache();

File: menubar/Sources/Pulse/CredentialCache.swift
import Foundation

/// Caches the access token so the secret store (macOS keychain / credentials
/// file) is read at most once per change. Every getToken() call does a
/// prompt-free fingerprint read; the secret is re-read only when the
/// fingerprint changes, so keychain ACL prompts are capped at roughly one per
/// Claude Code token rotation instead of one per poll.
final class CredentialCache {
    private let envToken: () -> String?
    private let readFingerprintFn: () -> String?
    private let readSecretFn: () throws -> String

    private var token: String?
    private var fingerprint: String?
    private var rejectedFingerprint: String?
    private var hasRejected = false
    private var rejectionError: Error = UsageError.auth

    init(
        envToken: @escaping () -> String? = { ProcessInfo.processInfo.environment["CLAUDE_USAGE_TOKEN"] },
        readFingerprintFn: @escaping () -> String? = readFingerprint,
        readSecretFn: @escaping () throws -> String = readAccessToken
    ) {
        self.envToken = envToken
        self.readFingerprintFn = readFingerprintFn
        self.readSecretFn = readSecretFn
    }

    /// Return the current access token.
    /// force bypasses only a previous rejection (an explicit user retry may
    /// prompt once); a healthy cached token is returned without any secret read.
    func getToken(force: Bool = false) throws -> String {
        if let env = envToken(), !env.isEmpty {
            return env
        }

        let fp = readFingerprintFn()

        if let token, fp == fingerprint {
            return token
        }

        guard let fp else {
            token = nil
            fingerprint = nil
            throw CredentialsError.notFound("Could not read credentials. Log in with Claude Code.")
        }

        if !force, hasRejected, fp == rejectedFingerprint {
            throw rejectionError
        }

        do {
            let secret = try readSecretFn()
            token = secret
            fingerprint = fp
            hasRejected = false
            return secret
        } catch {
            token = nil
            rejectedFingerprint = fp
            hasRejected = true
            let denied = CredentialsError.denied("Keychain access denied. Use Refresh Now to try again.")
            rejectionError = denied
            throw denied
        }
    }

    /// Call on HTTP 401/403: the cached token was rejected by the API.
    func invalidate() {
        token = nil
        rejectedFingerprint = fingerprint
        hasRejected = true
        rejectionError = UsageError.auth
    }
}
